patches/tModLoader/Terraria/ModLoader/TileLoader.cs

```diff
@@ -485,29 +499,97 @@ public static void DropCritterChance(int i, int j, int type, ref int wormChance,
+		bool isLarge = false;
+		if (Main.tileFrameImportant[type]) {
+			var tileData = TileObjectData.GetTileData(type, 0);
+			if (tileData != null) {
+				if(tileData.Width != 1 || tileData.Height != 1)
+					isLarge = true;
 			}
+			else if (TileID.Sets.IsMultitile[type])
+				isLarge = true;
 		}
+		if (!includeLargeObjectDrops && isLarge)
+			return true;
 
+		Tile t = Main.tile[i, j];
+		// Comment out to debug: Main.NewText($"Drop: {i}, {j}, {type}, L: {includeLargeObjectDrops}, HasTile: {t.HasTile}, type: {t.TileType}, fX: {t.TileFrameX}, name: {TileID.Search.GetName(t.TileType)}");
 		ModTile modTile = GetTile(type);
+		bool dropItem = modTile?.CanDrop(i, j) ?? true;
+		foreach (var hook in HookCanDrop) {
+			dropItem &= hook(i, j, type);
+		}
+		if (!dropItem)
+			return false;
 
-		if (modTile != null) {
-			if (!modTile.Drop(i, j)) {
-				return false;
-			}
+		foreach (var hook in HookDrop) {
+			hook(i, j, type);
+		}
+
+		return true;
+	}
 
-			if (modTile.ItemDrop > 0) {
-				Item.NewItem(new EntitySource_TileBreak(i, j), i * 16, j * 16, 16, 16, modTile.ItemDrop, 1, false, -1);
+	public static void GetItemDrops(int x, int y, Tile tileCache, ref int dropItem, ref int dropItemStack, bool includeLargeObjectDrops = false, bool includeAllModdedLargeObjectDrops = false)
+	{
+		ModTile modTile = GetTile(tileCache.TileType);
+		if (modTile == null)
+			return;
+
+		// Various call sites to WorldGen.KillTile_DropItems expect different sets of tile drops to be retrieved:
+		// KillTile: All 1x1 tiles
+		// ReplaceTile: All 1x1 tiles, all supported multitiles
+		// CheckModTile: All modded tiles (except 1x1 tiles will drop from killtile)
+		bool needDrops = false;
+		TileObjectData tileData = TileObjectData.GetTileData(tileCache.TileType, 0, 0);
+		if (tileData == null) {
+			// Terrain tile
+			needDrops = true;
+		}
+		else if (tileData.Width == 1 && tileData.Height == 1) {
+			// 1x1 tile, includeAllModdedLargeObjectDrops prevents double spawns from framing code calling CheckModTile, which calls KillTile_DropItems and KillTile. (Bars)
+			needDrops = !includeAllModdedLargeObjectDrops;
+		}
+		else if (includeAllModdedLargeObjectDrops)
+			needDrops = true;
+		else if (includeLargeObjectDrops) {
+			if (TileID.Sets.BasicChest[tileCache.type] || TileID.Sets.BasicDresser[tileCache.type]) {
+				needDrops = true;
 			}
+		}
+		if (!needDrops) {
+			return;
+		}
 
-			return false;
+		var itemDrops = modTile.GetItemDrops(x, y);
+		foreach (var item in itemDrops) {
+			item.Prefix(-1); // Assign a random prefix, as expected
+			int num = Item.NewItem(WorldGen.GetItemSource_FromTileBreak(x, y), x * 16, y * 16, 16, 16, item, noBroadcast: false);
+			Main.item[num].TryCombiningIntoNearbyItems(num);
 		}
+	}
 
-		return true;
+	public static int GetItemDropFromTypeAndStyle(int type, int style = 0, bool allowFallbackToStyleZero = true)
```

Comment:
Is this ever called with `allowFallbackToStyleZero: false`? Is this just intended for modders?

Reply:
Yeah, just there in case it comes up.

Reply:
I can't really think of a use, especially because you still allow `ItemDrop` override. Prefer to add on request.